src/drivers/magnetometer/bosch/bmm150/BMM150.cpp

```diff
@@ -0,0 +1,473 @@
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "BMM150.hpp"
+
+using namespace time_literals;
+
+static constexpr int16_t combine(uint8_t msb, uint8_t lsb)
+{
+	return (msb << 8u) | lsb;
+}
+
+BMM150::BMM150(I2CSPIBusOption bus_option, int bus, int bus_frequency, enum Rotation rotation) :
+	I2C(DRV_MAG_DEVTYPE_BMM150, MODULE_NAME, bus, I2C_ADDRESS_DEFAULT, bus_frequency),
+	I2CSPIDriver(MODULE_NAME, px4::device_bus_to_wq(get_device_id()), bus_option, bus),
+	_px4_mag(get_device_id(), rotation)
+{
+	_px4_mag.set_external(external());
+}
+
+BMM150::~BMM150()
+{
+	perf_free(_reset_perf);
+	perf_free(_bad_register_perf);
+	perf_free(_bad_transfer_perf);
+	perf_free(_overflow_perf);
+}
+
+int BMM150::init()
+{
+	int ret = I2C::init();
+
+	if (ret != PX4_OK) {
+		DEVICE_DEBUG("I2C::init failed (%i)", ret);
+		return ret;
+	}
+
+	return Reset() ? 0 : -1;
+}
+
+bool BMM150::Reset()
+{
+	RegisterWrite(Register::POWER_CONTROL, 0);
+	_state = STATE::RESET;
+	ScheduleClear();
+	ScheduleDelayed(1_ms);
+	return true;
+}
+
+void BMM150::print_status()
+{
+	I2CSPIDriverBase::print_status();
+
+	perf_print_counter(_reset_perf);
+	perf_print_counter(_bad_register_perf);
+	perf_print_counter(_bad_transfer_perf);
+	perf_print_counter(_overflow_perf);
+}
+
+int BMM150::probe()
+{
+	const uint8_t POWER_CONTROL = RegisterRead(Register::POWER_CONTROL);
+	const uint8_t CHIP_ID = RegisterRead(Register::CHIP_ID);
+
+	PX4_DEBUG("POWER_CONTROL: 0x%02hhX, CHIP_ID: 0x%02hhX", POWER_CONTROL, CHIP_ID);
+
+	// either power control bit is set and chip ID can be read, or both registers are 0x00
+	if ((POWER_CONTROL & POWER_CONTROL_BIT::PowerControl) && (CHIP_ID == chip_identification_number)) {
+		return PX4_OK;
+
+	} else if ((POWER_CONTROL == 0) && (CHIP_ID == 0)) {
+		return PX4_OK;
+	}
+
+	return PX4_ERROR;
+}
+
+float BMM150::compensate_x(int16_t mag_data_x, uint16_t data_rhall)
+{
+	float retval = 0;
+
+	// Processing compensation equations
+	//  not documented, but derived from https://github.com/BoschSensortec/BMM150-Sensor-API/blob/a20641f216057f0c54de115fe81b57368e119c01/bmm150.c#L1624-L1633 as of 2020-09-25
+	float process_comp_x0 = (((float)_trim_data.dig_xyz1) * 16384.f / data_rhall);
```

Comment:
What do you think about changing all of the c-style casts to static_casts?

Reply:
I was trying to keep this (somewhat hideous) code in the same form as the official driver from Bosch.